File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, g: List[List[int]]) -> int:
        
        n = len(g)
        m = len(g[0])
        
        
        if g[0][0]==1 or g[n-1][m-1]==1:
            return 0
        
        dp = arr = [[-1 for j in range(m)] for i in range(n)]
        
        def help(r,c):
            if(r==n-1) and (c ==m-1):
                return 1
            
            if (r < 0) or (c < 0) or (r>=n) or (c>=m) or (g[r][c]==1):
                return 0
            
            if dp[r][c] != -1:
                return dp[r][c]
            
            # g[r][c] =1
            
            pathCount = help(r+1, c) + help(r,c+1)
            
            # g[r][c] = 0
            
            dp[r][c] = pathCount
            
            return dp[r][c]
        
        return help(0,0)

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    // int ans;
    int n, m;
    vector<vector<int>> dp;
    int help(vector<vector<int>>& g, int r, int c){
        if(r==n-1 and c==m-1){
            // ans = ans+1;
            return 1;
        }
        
        if(r < 0 || c < 0 || r>=n || c>=m|| g[r][c] == 1)
            return 0;
        
        if(dp[r][c] !=-1)
            return dp[r][c];
        
        g[r][c] =1;
        int currPath = help(g, r+1,c) + help(g, r, c+1);
        g[r][c] =0;
        
        dp[r][c] = currPath;
        return dp[r][c];
    }
    int uniquePathsWithObstacles(vector<vector<int>>& g) {
        n = g.size();
        m = g[0].size();
        if(g[0][0] || g[n-1][m-1])
            return 0;
        
        dp.resize(n, vector<int>(m,-1));
        int ans = help(g, 0, 0);
        return ans;
        
    }
};